### ISSUE TITLE
Breaking Change in Vec::truncate drop order between 1.40 -> 1.41

### ISSUE THREAD

**Issue by mgattozzi** [rust-lang/rust#68709]
Context:
[The commit that introduced the change.](https://github.com/rust-lang/rust/commit/9e8c4e6fb1c952048fb823e59f4c9c6487bf9a58)
[The PR for the commit](https://github.com/rust-lang/rust/pull/64432)

Minimal repro can be found in the commit and the PR. At work we have a private internal crate that depended on this drop ordering. As we had not run the beta train we did not catch it until attempting to upgrade our stable version of rust to 1.41.0. Now while the behavior of the drop order was unspecified it has been stabilized.

I'm sure we can find a workaround for this at work, but I'm filing this issue as it brings up some interesting questions that I don't know if the libs team has ever taken an official stance on.

If the observable behavior changes but isn't specified is this a breaking change?
The API itself has not changed, but the drop order has changed. I'm of the opinion it is a breaking change even if it brings the API more in line with how slice drop order works, though talking to some friends they are of the opinion it is not a breaking change.

Anyways I wanted to file this issue to bring attention to this and maybe get a more clear ruling from the libs team on what they consider is and is not breaking change behavior. If more info is needed on why we needed this @estebank and I can probably give more specifics on this, though I don't think this is so much trying to fix a bug as it is determining whether a change should be reverted.

Thanks! 😄

**Comment by sfackler** [comment]
> I'm sure we can find a workaround for this at work

```rust
fn truncate_with_old_drop_order<T>(vec: &mut Vec<T>, len: usize) {
    while vec.len() > len {
        vec.pop();
    }
}
```

**Comment by estebank** [comment]
@sfackler we can deal with the fallout ourselves, but the fact that we were hit by this means that _others_ in the ecosystem that are also not open source might have been hit by this change in behavior. The question is whether that is acceptable breakage.

**Comment by Lucretiel** [comment]
Seem like it is [specified] in the documentation for `Vec`, if not the specific method:

> Vec does not currently guarantee the order in which elements are dropped. The order has changed in the past and may change again.

**Comment by the8472** [comment]
[RFC 1105](https://rust-lang.github.io/rfcs/1105-api-evolution.html#behavioral-changes)

> This RFC does not attempt to provide a comprehensive policy on behavioral changes, which would be extremely difficult. In general, APIs are expected to provide explicit contracts for their behavior via documentation, __and behavior that is not part of this contract is permitted to change in minor revisions.__ (Remember: this RFC is about setting a minimum bar for when major version bumps are required.)

[emphasis mine]

**Comment by Ixrec** [comment]
Also relevant is [RFC 1857: Stabilize drop order](https://github.com/rust-lang/rfcs/pull/1857), which in retrospect is oddly unclear about precisely how this applies to library types. The RFC text [explicitly states a drop order for `Vec`](https://github.com/rust-lang/rfcs/blob/master/text/1857-stabilize-drop-order.md#slices-and-vec), but with [an exemption for certain corner cases](https://github.com/rust-lang/rfcs/blob/master/text/1857-stabilize-drop-order.md#allowed-unspecified-behavior). Then Niko said ["I hadn't thought of containers as being in the scope of this RFC per se ... Interestingly, only Vec seems to have a drop-order that is particularly aligned with "creation order"."](https://github.com/rust-lang/rfcs/pull/1857#issuecomment-281408206) But several later comments in the thread by Niko and others ([1](https://github.com/rust-lang/rfcs/pull/1857#issuecomment-282210600), [2](https://github.com/rust-lang/rfcs/pull/1857#issuecomment-282839422), [3](https://github.com/rust-lang/rfcs/pull/1857#issuecomment-284550843), [4](https://github.com/rust-lang/rfcs/pull/1857#issuecomment-285406019), [5](https://github.com/rust-lang/rfcs/pull/1857#issuecomment-286264016), etc) are clearly discussing `Vec`, sometimes even indicating they're treating `Vec`s and primitive arrays the same. I can't seem to find a single comment implying, much less arguing, that `Vec` should be treated differently from the primitive types.

However, AFAICT not a single comment in that thread ever raised the question of `Vec` methods which drop elements of the `Vec`, other than the regular `Drop` impl. Thus, my interpretation is that RFC 1857 only applies to the `Drop` trait `impl` of standard library types. So I believe this specific question about `Vec::truncate()`'s behavior would fall under the RFC 1105 policy quoted above, i.e. it's unspecified and is allowed to change. However, that also means `Vec`'s documentation stating ["Vec does not currently guarantee the order in which elements are dropped. The order has changed in the past and may change again."](https://doc.rust-lang.org/std/vec/struct.Vec.html#guarantees) is simply incorrect and outdated. I'm not sure there is a plausible interpretation of RFC 1857 that doesn't contract that documentation.

-----

Language-lawyering aside, this is definitely more subtle and unclear than it ought to be, and there's room for improvement. At the bare minimum, we need to sort out the apparent contradiction between Vec's current documentation and RFC 1857, and generally make our documentation around drops more explicit about when we mean a type's `Drop` impl versus a method that leads to dropping.

Because `std` container types already have exceptionally strong stability commitments, even by Rust standards (e.g. [`Vec`'s layout guarantees](https://doc.rust-lang.org/std/vec/struct.Vec.html#guarantees)), it seems reasonable to "just define all the behavior" in this case. In other words, I currently think we should (eventually, not urgently) come up with a list of all "non-Drop drops" in the standard library, then commit to and document a specific drop order for each of them, unless there's some deep reason why that can't work (maybe we'd make an exception for iterator types?).

**Comment by Lucretiel** [comment]
into_iter is arguably another one (though it's technically `IntoIter` doing the dropping)

**Comment by Lucretiel** [comment]
`clear` and `retain` are some others

**Comment by nikomatsakis** [comment]
@Ixrec With regard to RFC 1857, quite frankly, those comments may have been in ignorance of the documentation of `Vec`. I guess I am surprised to learn that we had not guaranteed the drop order for `Vec`. But that RFC was (I believe) primarily discussed by folks from lang team, not libs team, so maybe we all just neglected to read the docs. =)

**Comment by spastorino** [comment]
Going to remove `I-prioritize` because this is not `T-compiler` nor `libs-impl`.

**Comment by nikomatsakis** [comment]
So...it's been 5 months here. I think it's unlikely that we're going to take any action to further change the drop order of `Vec::truncate`. Further, Vec's documentation does state:

> Vec does not currently guarantee the order in which elements are dropped. The order has changed in the past and may change again.

Therefore, I'm going to close this issue. Thanks to @mgattozzi for raising it. Feel free to re-open if you think this is unjustified.

**Comment by mgattozzi** [comment]
No I think this is a good enough reason and we only had tests relying on it that we changed. Since this has never been a guarantee I think this is a good reason to close 😁👍🎉
